### ISSUE TITLE
Single-line analysis routines

### ISSUE THREAD

**Issue by mdeceglie** [NREL/rdtools#36]
As a longer-term goal, we would like to develop analysis routines that bundle normalization, filtering, aggregation, and analysis so they can be deployed with a single line.

**Comment by mdeceglie** [comment]
We've now considered several different ways to accomplish this (#116 #117 #130 ). Let's use this issue to facilitate discussion of the best way forward. Of particular interest is the question of whether to require that the single line analysis routine API can be programmed by a separate "recipe file". This would have a few benefits. First, it would allow us to decouple changes in analysis best practices from RdTools major version releases. Second, it would provide analysts and researchers a way to compactly share their routines built with RdTools. If we took this approach I'm imagining the part that gets incorporated and version controlled with RdTools is more of a shell, and the detailed plumbing of the entire routine gets programmed in through the recipe file. 

No matter what, we intend to build this on the functional interface, so people can use that as a library to meet their own needs. 

@cdeline @cwhanse @GregoryKimball @abshinn @kanderso-nrel What do you think about this point?

**Comment by cwhanse** [comment]
+1 to the "recipe file" concept. I'm not confident enough to say that one of #116, #117 or #130 is better. #130 adds a layer that is unfamiliar to me and thus I'm uncertain about being about to use, debug or contribute to that approach.

**Comment by kanderso-nrel** [comment]
[PVMismatch](https://github.com/SunPower/PVMismatch) uses a property-based approach in its [PVCell](https://github.com/SunPower/PVMismatch/blob/master/pvmismatch/pvmismatch_lib/pvcell.py) class that is similar in spirit to the declarative approach in #130.  Here's a toy example in the context of RdTools:

``` python
class SystemAnalysis:
    
    def __init__(self, poa, power):
        self.poa = poa
        self.power = power
    
    @property
    def normalized_power(self):
        return rdtools.normalize(self.power, self.poa)

    @property
    def filtered_power(self):
        return rdtools.filter(self.normalized_power)

    @property
    def degradation_result(self):
        results = rdtools.degradation(self.filtered_power)
        return results

poa, power = ...
sa = SystemAnalysis(poa, power)
results = sa.degradation_results
```
Using `@property` on a function lets you treat it like a normal static attribute, but really it's calling the function under the hood and giving you the returned value.  @mikofski since it looks like you set that up in https://github.com/SunPower/PVMismatch/commit/33189813f9033c360de308ce68ee9b795acf60e7, do you think it was a good idea or maybe it should have been done differently in hindsight?

**Comment by mikofski** [comment]
Considering the redundant calls to properties in PVMismatch, (see https://github.com/SunPower/PVMismatch/issues/121) I do regret it a little bit. For example look at [`self.Isat1`](https://github.com/SunPower/PVMismatch/blob/793f6fbc91f0a55cfec7f5861ce4e0acd8b8e5e1/pvmismatch/pvmismatch_lib/pvcell.py#L136-L146) which is then called 3 times in [`self.Voc`](https://github.com/SunPower/PVMismatch/blob/793f6fbc91f0a55cfec7f5861ce4e0acd8b8e5e1/pvmismatch/pvmismatch_lib/pvcell.py#L168-L178), itself a property and then 4 more times in [`_VocSTC`](https://github.com/SunPower/PVMismatch/blob/793f6fbc91f0a55cfec7f5861ce4e0acd8b8e5e1/pvmismatch/pvmismatch_lib/pvcell.py#L180-L196). However I think [the caching/memoization idea that Kevin suggested](https://github.com/kanderso-nrel/PVMismatch/blob/pvcell_cache/pvmismatch/pvmismatch_lib/pvcell.py) should resolve that issue.

In general I see Python properties and [Python descriptors](https://docs.python.org/3/howto/descriptor.html) in general as similar to [MATLAB depndent properties](https://www.mathworks.com/help/matlab/matlab_oop/access-methods-for-dependent-properties.html) they're only evaluated when necessary and therefore adapt to changes automatically. I think in Python they serve the additional purpose of making an immutable member, which I believe is impossible with Python attributes.

So to me Python properties just replaces an equivalent function call like `get_Isat1()` using the example from above, which might make it more explicitly a function call and therefore encourage the user to locally cache the value like `Isat1 = self.get_Isat1()`.

But in the context of this issue here regarding pipeline analysis, I don't immediately see the usage pattern for Python properties. Sorry for being obtuse, but I would need an example to see how it would compare to the PR's already mentioned.

---

Weighing in on this issue, I think I either on of #117 or #130 may work, with #117 being slightly easier but possible more rigid, and #130 more general, but slightly more work initially. I have a very slight preference for #130 having promoted somthing very similar for many years (anyone remember [Carousel](https://breakingbytes.github.io/simkit/)?). But honestly I think it's too early to embark on either of these while the API of rdtools in general seems to be undocumented. Before proceeding with this issue, I think rdtools should publish the sphinx docs either as ghpages or readthedocs, asap.

Model chains (#117) seems to take the approach that there pre-processes and analyses, and that a matrix of these form independent results. (See [`SystemAnalysis_Example` notebook](../blob/36ca2bd6a19a263d6b008b3ce2e354a8f208bc7f/docs/SystemAnalysis_example.ipynb) As long as this is true, then this approach is probably the easiest. So for example one could implement a new analysis or a new pre-process, and therefore implement a new result. This would require the API for pre-process to be standardized and published so that it could be utilized by any analysis, and ditto for the analyses, define a standard API. Questions/task would be can the API be defined? Show an example of how rdtools would be extende to include a new analysis or pre-process.

**Comment by cwhanse** [comment]
I'm not yet certain that this is an actual issue, but, I find the dynamic properties to get in the way when diagnosing issues with calculations. To me, they seem a bit like Schrodinger's cat: how can I be confident that the act of inspecting the property hasn't altered it's state, or, how can I know when the state has changed?

I've been trying to reproduce PVMismatch calculations using pvlib functions, and I can't yet get the two calculations to match The disagreement is in the short circuit current. Two possible causes, which may both in involved:

-  PVMismatch formulates the diode models differently, where the input parameter is short circuit current at reference conditions `Isc0_T0` rather than photocurrent at reference conditions
- Some lack of synchronizing among the properties, which could be a user error rather than a code error.

When I have code that reproduces the problem, I'll post to PVMismatch.

**Comment by kanderso-nrel** [comment]
Thanks for the thoughtful comments.  

> (Mark) Before proceeding with this issue, I think rdtools should publish the sphinx docs either as ghpages or readthedocs, asap.

ICYMI -- documentation is one of the focuses for RdTools 2.0 and the results of merging a few sphinx-related PRs to the development branch are available here: https://rdtools.readthedocs.io

We haven't really publicized those docs because master and the latest release aren't sphinx-compatible, so the docs in that link don't apply to what people will have installed from PyPI.  We are planning on combining the new docs, the new functions (e.g. the soiling SRR code), and whatever we decide on for this single-line analysis API into one big major release.  Do you see an advantage to releasing the docs officially before focusing on this feature, or is it just a matter of making sure documentation makes it into the next release (which I agree is high priority and is well underway)?  

> (Mark) regarding model-chains, I don't immediately see the usage pattern for Python properties. 

What I was imagining is a sort of compromise between #117 and #130.  The overall [RdTools calculation graph](https://rdtools.readthedocs.io/en/model_plugins/_images/model_graph_diagram.png) is fairly complex, so I like the idea of lots of small bits of code that are individually only responsible for a single node in the graph.  The pattern here would be to use one property per node.  For instance the cell temperature property would figure out whether it can use module temperature or if it needs to fall back to modeling with ambient temperature.  Another thing -- I'm not sure it's ever been stated on github, but one desired feature of this high-level API is that it would store all the intermediate variables for later inspection.  I see (cached) properties as a way to provide standard/pythonic access to the intermediate variables while keeping the under-the-hood calculation structure general and not needlessly coupled to our existing analysis workflows.  TL;DR: more approachable than #130, more flexible than #117.

> (Cliff) I find the dynamic properties to get in the way when diagnosing issues with calculations. To me, they seem a bit like Schrodinger's cat: how can I be confident that the act of inspecting the property hasn't altered it's state, or, how can I know when the state has changed?

I could be convinced that the "magic" under the hood here (and in #130) aren't worth it and we should just keep it simple.  My main resistance to giving up flexibility comes from uncertainty about future expansion -- I'd be fine with the straightforward approach in #117 if we keep the scope to just soiling and degradation instead of trying to leave the door open for any and all future system analyses.

**Comment by mikofski** [comment]
>Do you see an advantage to releasing the docs officially before focusing on this feature, or is it just a matter of making sure documentation makes it into the next release (which I agree is high priority and is well underway)?

Yes, I think the docs should be the number one priority. To me the single-line analysis feature is a noble high-level goal, but in my opinion, the documentation and the low-level functions are the foundations of rdtools, while the single-line analysis is a high-level abstraction that I believe will be difficult to implement without a published API. In general I think the documentation and tests should always come first. In my opinion, publishing documentation forces you to accept or create an API.

>We are planning on combining the new docs, the new functions (e.g. the soiling SRR code), and whatever we decide on for this single-line analysis API into one big major release.

I believe that a big major release is more challenging and more prone to problems that lots of incremental tiny releases, starting of course with unit testing and documentation, followed by functions to implement the desired API.

>We haven't really publicized those docs because master and the latest release aren't sphinx-compatible, so the docs in that link don't apply to what people will have installed from PyPI.

I believe part of the problem is that the Rdtools workflow is unnecessarily complicated by having multiple release branches. In particular, I don't think a development branch is necessary in a workflow that includes continuous integration. In my opinion, the best workflow is the "feature branch workflow" which I wrote about in this post called ["the winning workflow"](https://poquitopicante.blogspot.com/2016/10/winning-workflow.html). The main benefit of the feature branch workflow is PRs are branched off from a single release branch called master. Everyone works out of master, it is always ready to deploy, and it represents the bleeding edge. There are no other branches, and feature PR's are deleted after they're merged.

**Comment by mdeceglie** [comment]
We have in the past discussed releasing 2.0 without the single line analysis routines. This is the best path to getting the docs aligned with master/release in the near term. Especially without a clear path that everyone is happy with on the single-line routines, I would support saving the additive functionality of the single-line routines for 2.x. Note that because the development branch includes #105, a major version release is warranted.

@cdeline  what do you think?

**Comment by mdeceglie** [comment]
I think @mikofski  hits the nail on the head regarding a big drawback of the approach of #117 

> I think it would help to show an example of how rdtools would be extended to include a new analysis or pre-process.

The approach of #117 is at risk of becoming unmaintainable as we keep adding new analyses.

**Comment by cdeline** [comment]
I'm going to weigh in here because there are a lot of good suggestions in this thread.   Based on Mark's great [blog post](https://poquitopicante.blogspot.com/2016/10/winning-workflow.html), there are three major things that we should be doing immediately:  
1. Feature branches exist in personal forks and pull requests reference these.  This avoids our current situation which is an impossibly tangled remote network graph.   Check pvlib's network for a good aspirational example.
2. incremental builds and very frequent updates to Master.  It's been AGES since we've had a numbered release, but there's no reason we can't have patches and nightly builds, then do a major tagged release quarterly or so.
3. Start with tests and pull requests before you write code: I feel less strongly about this one, but if you want to go this route, ok with me.  Maybe wait to assign reviewers until you're ready to have someone take a look.
@mdeceglie , @kanderso-nrel - give a thumbs up if you agree that we should tidy up our feature branches and banish them to personal forks.

**Comment by cdeline** [comment]
As for the other details in this thread, I agree that a 2.0 release with Sphinx docs and updated API is in order, followed by a 2.1 release with our alpha #117 model chains since that is what is currently used in our Fleets pipeline.  I'm less concerned about hypothetical future cases in which new analyses make our workflow un-maintainable.  I'm happy with keeping solutions such as #130 in side development forks until they're more fully baked and the complexity downside is warranted.

**Comment by mdeceglie** [comment]
I am on board with changing the github workflow, but with the caveat that we make that change after we get the current `development` changes into a release (2.0). Once that is cut, then we can clean up and adopt the new workflow.

We all agree to hold off on the single-line routines until 2.x (x>0), so I don't foresee any showstoppers for 2.0.

**Comment by mdeceglie** [comment]
Closed with #289
